### PR TITLE
persist: consolidate as we go in snapshot_and_fetch

### PR DESCRIPTION
**Pull request by danhhz**
Sometimes a snapshot will be read which consolidates really well (e.g. when the since has been heavily held back but we're reading near the upper). Doing something reasonable here in `persist_source` is complicated and tracked in https://github.com/MaterializeInc/materialize/issues/16858, but for `snapshot_and_fetch` (which is used in the controllers) it's straightforward to consolidate as we go to keep memory usage down.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Tips for reviewer

The first commit is just some code movement. Look at the second commit to easily see the actual diff.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by danhhz** [comment]
@frankmcsherry Marked as a draft because I heard the idea second-hand and am a little bit just guessing. Please nitpick what's going on here!

File: src/persist-client/src/read.rs [hunk 1]
```diff
@@ -881,6 +858,53 @@ where
...
+    K: Debug + Codec + Ord,
+    V: Debug + Codec + Ord,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64 + Send + Sync,
+{
+    /// Generates a [Self::snapshot], and fetches all of the batches it
+    /// contains.
+    ///
+    /// To keep memory usage down when reading a snapshot that consolidates
+    /// well, this consolidates as it goes.
```
Comment (benesch) [inline src/persist-client/src/read.rs:872]: I think either add a note that the final result may yet _not_ be consolidated—or slap a consolidate on the end?

**Review by benesch — COMMENTED** [review]
Thanks for hopping on this one so quickly! Gonna do wonders for envd robustness in the face of slow compaction.

LGTM but leaving to @frankmcsherry to approve.

File: src/persist-client/src/read.rs (hunk 1 above)
Comment (benesch) [reply src/persist-client/src/read.rs:872]: > Also curious if you think we should put a consolidate at the end and then document that the output is guaranteed to be consolidated

Missed this in your PR description, oops. 🤦🏽‍♂️ If we decide not to consolidate, I do think it's worth making the docstring mention that we're not consolidating a bit more explicitly.

**Review by frankmcsherry — APPROVED** [review]
This looks good to me. I had some bonus thoughts that shouldn't hold up this work (it seems a strict improvement):
1. It is plausible that one eventually wants a `F: Fn(K,V) -> (K,V)` argument, which looks like an MFP you might be pushing down. Reason being, ofc, that if you are projecting or transforming in a way that allows further consolidation, amazing. Ofc also .. it probably has to produce `Result` types because errors and maybe it is horrible. But it is an opportunity to squish things even more if we know they are *about* to be dramatically reduced.
2. Do the parts themselves have an order? In DD at least they are stored by `(K, V, T)` and so all of the compaction that can occur is within a `(K, V)` pair and they are just all pre-bundled. You apply the frontier stuff to `T` and then do a consolidate on the times. Still good to consolidate across batches of course.
3. Agree with @benesch that knowing if it is for sure entirely consolidated.
4. I'm not sure I've gotten my head around how `persist_source` works, in that I thought we applied MFP stuff there and then consolidated, but if we are using this function (maybe we are not) then the source wouldn't be as effective as I hoped (loading up stuff before applying the MFP). Locally this looks like a strict improvement, so I'm probably misremembering how `persist_source` works, but also my takes on e.g. MFP stuff might just be wrong if this method is not used by any code like that.

**EDIT**: Bonus thought: 
5. If the batches are themselves sorted, another strategy is to perform a merge as they arrive rather than maintain a buffer. This would allow you to stream out the results (potentially to MFP applying operators, or monotonic-friendly operators). This would have the potential for a smaller memory footprint (holding on to the heads of each of the runs of parts).

I think many of the thoughts are of the form: "this looks great, and I want to help nudge things in this space further" :D

**Comment by danhhz** [comment]
> This looks good to me. I had some bonus thoughts that shouldn't hold up this work (it seems a strict improvement):
> 
> 1. It is plausible that one eventually wants a `F: Fn(K,V) -> (K,V)` argument, which looks like an MFP you might be pushing down. Reason being, ofc, that if you are projecting or transforming in a way that allows further consolidation, amazing. Ofc also .. it probably has to produce `Result` types because errors and maybe it is horrible. But it is an opportunity to squish things even more if we know they are _about_ to be dramatically reduced.
> 2. Do the parts themselves have an order? In DD at least they are stored by `(K, V, T)` and so all of the compaction that can occur is within a `(K, V)` pair and they are just all pre-bundled. You apply the frontier stuff to `T` and then do a consolidate on the times. Still good to consolidate across batches of course.
> 4. I'm not sure I've gotten my head around how `persist_source` works, in that I thought we applied MFP stuff there and then consolidated, but if we are using this function (maybe we are not) then the source wouldn't be as effective as I hoped (loading up stuff before applying the MFP). Locally this looks like a strict improvement, so I'm probably misremembering how `persist_source` works, but also my takes on e.g. MFP stuff might just be wrong if this method is not used by any code like that.
> 
> **EDIT**: Bonus thought: 5. If the batches are themselves sorted, another strategy is to perform a merge as they arrive rather than maintain a buffer. This would allow you to stream out the results (potentially to MFP applying operators, or monotonic-friendly operators). This would have the potential for a smaller memory footprint (holding on to the heads of each of the runs of parts).
> 
> I think many of the thoughts are of the form: "this looks great, and I want to help nudge things in this space further" :D

Yeah, this indeed is **not** used by persist_source! They share a huge chunk of code, but this is essentially a convenience method for users (storage controller, reclocking, test code) that are running outside of a dataflow and also believe that the consolidated size of the shard is small.

These are great thoughts, though! We should be able to reuse whatever we do for consolidate on read in persist_source here as well (cc @bkirwi). Went ahead and wrote them all down here as opportunities for future improvements.

> I think either add a note that the final result may yet not be consolidated—or slap a consolidate on the end?

> 3. Agree with @benesch that knowing if it is for sure entirely consolidated.

Audited all non-test callers of this method. None of them call consolidate, but at least some (all?) would benefit from consolidation, so it seems to be a reasonably common mistake to assume that this method returns consolidated output. Went ahead and just added that to the contract of the method.

**Comment by bkirwi** [comment]
Yep, agree with Dan!

As it happens the `ConsolidateBuffer` in the persist source already means the output is consolidated as we go, so everything we're looking at in https://github.com/MaterializeInc/materialize/issues/16858 is dealing with additional constraints in the source: more data, multiple workers, ~smooth output.

**Comment by danhhz** [comment]
TFTRs!!

**Pull request merged**